connection_pool.go

```diff
@@ -66,46 +68,94 @@ func (b *backend) GetConnection() *connection {
 	}
 }
 
+// debugHitMaxConnections will potentially do some debugging output to help diagnose situations
+// where we're hitting connection limits.
+func (b *backend) debugHitMaxConnections() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if time.Now().Before(b.debugTime) {
+		return
+	}
+	b.debugTime = time.Now().Add(10 * time.Second)
+
+	log.Warn("DEBUG: hit max connections",
+		"counter", atomic.LoadInt32(b.counter),
+		"len(conns)", len(b.conns))
+	for idx, conn := range b.conns {
+		log.Warn("DEBUG", "connection",
+			"idx", idx,
+			"conn", conn,
+			"closed", conn.IsClosed(),
+			"age", time.Now().Sub(conn.StartTime()))
+	}
+}
+
 // getNewConnection establishes a new connection if and only if we haven't hit the limit, else
 // it will return nil. If an error is returned, we failed to connect to the server and should
 // abort the flow.
 func (b *backend) getNewConnection() (*connection, error) {
-	for {
-		if ctr := b.NumOpenConnections(); int(ctr) >= b.conf.ConnectionLimit {
-			return nil, nil
-		} else {
-			// Now attempt to increment and swap to ensure we don't race.
-			if !atomic.CompareAndSwapInt32(b.counter, int32(ctr), int32(ctr)+1) {
-				log.Debug("failed counter race", "addr", b.addr)
-				continue
-			}
+	if ctr := b.NumOpenConnections(); int(ctr) >= b.conf.ConnectionLimit {
+		go b.debugHitMaxConnections()
+		return nil, nil
+	} else {
```

Comment:
If the problem is that created connections can be lost / permanently count against the limit despite being dead, then this won't help. It would just mean we will settle down at a slightly larger but still finite number of dead connections.